text/0000-really-tagged-unions.md

```diff
@@ -0,0 +1,362 @@
+
+We may need to double-check that the sub-structs are actually being marked as `repr(C)`.
+
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There aren't really any drawbacks: we already decided to do this. This is just making it official.
+
+
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+There are a few alternative interpretations of `repr(X)` on a non-C-like enum.
+
+
+
+## It should do nothing 
+
+In which case it should probably become an error/warning. This isn't particularly desirable, as was discussed [when we decided to maintain this behaviour](https://github.com/rust-lang/rust/issues/40029).
+
+
+
+## The tag should come after the union, and/or order should be manually specified
+
+There isn't a particularly compelling reason to move the tag around because of how padding and alignment are handled: you can't actually save space by putting the tag after, as long as your tag is a reasonable size.
+
+It's possible positioning the tag afterwards could be desirable to interoperate with a definition that is provided by a third party (hardware spec or some existing C library). However there are tons of other tag packing strategies that we also can't handle, so we'd probably want a more robust solution for those kinds of cases anyway.
+
+
+
+## Compound variants shouldn't automatically be marked as `repr(C)`
+
+It's possible in the pure-Rust case to only want the outer-most level to be well-defined and to still let the inner contents be aggressively optimized. This interpretation and the RFC's primary interpretation can always be implemented on top of the other one by just making all compound variants into outline structs, just like how the desugarrings in the previous sections were written.
+
+So this is just a matter of "what is a good default". The FFI case clearly wants fully defined layouts, while the pure-Rust case seems like a toss up. It seems like `repr(C)` is therefore the more ergonomic default.
+
+
+## The tag shouldn't be opaque to outer enums
+
+So for instance `Option<MyEnum>` could be the same size as `MyEnum`. The author hasn't fully thought this issue out. It seems like it might be fine? It just seems safest to completely abandon optimization. Feedback is welcome here.
+
+
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+The unresolved questions basically amount to future extensions of this design. Here are some basic sketches/notes for things which are possible, but which have been left out for simplicity.
+
+* A field/method for the tag/payload (my_enum.tag, my_enum.payload)
+    * Probably should be a field to avoid conflicts with user-defined methods
+    * Might need `#[repr(pub(X))]` for API design reasons
+* Compiler-generated definitions for the Repr types
+    * With inherent type aliases on the enum? (`MyEnum::Tag`, `MyEnum::Payload`, `MyEnum::PayloadA`, etc.)
```

Comment:
I think that these should probably be implemented separate from this RFC. I'd be inclined to instead allow the ecosystem to generate these declarations using a Custom Derive like `#[derive(EnumRepr)]` which would define an `my_enum_repr` (or similar) module with `Tag`, `Payload`, and structs for each payload. 

Right now I don't think we have the tools to expose these in a nice way, especially if we consider a theoretical enum with a variant called `Tag` or `Payload` (this change would be breaking for those enums).

Reply:
Yes, I was willing to punt on these since serde and cbindgen could both generate these automatically if need be.

Reply:
This sounds partly like #1450.